osu.Game/Rulesets/UI/RulesetInputManager.cs

```diff
@@ -153,10 +163,55 @@ protected virtual RulesetKeyBindingContainer CreateKeyBindingContainer(RulesetIn
 
         public class RulesetKeyBindingContainer : DatabasedKeyBindingContainer<T>
         {
+            public event Func<T?, List<T>, bool> HandleBindings;
+            private UIEvent lastBlocked;
+
             public RulesetKeyBindingContainer(RulesetInfo ruleset, int variant, SimultaneousBindingMode unique)
                 : base(ruleset, variant, unique)
             {
             }
+
+            protected override bool Handle(UIEvent e)
+            {
+                if (!(e is KeyDownEvent || e is MouseDownEvent || e is JoystickButtonEvent)) return base.Handle(e);
+                if (e is KeyDownEvent ev && ev.Repeat) return base.Handle(e);
+
+                var pressedCombination = KeyCombination.FromInputState(e.CurrentState);
+                var combos = KeyBindings?.ToList().FindAll(m => m.KeyCombination.IsPressed(pressedCombination, KeyCombinationMatchingMode.Any));
+
+                if (combos != null)
+                {
+                    var pressedActions = combos.Select(c => c.GetAction<T>()).ToList();
+
+                    InputKey? key = null;
+                    if (e is KeyDownEvent kb)
+                        key = KeyCombination.FromKey(kb.Key);
+                    else if (e is MouseDownEvent mouse)
+                        key = KeyCombination.FromMouseButton(mouse.Button);
+                    else if (e is JoystickButtonEvent joystick)
+                        key = KeyCombination.FromJoystickButton(joystick.Button);
+
+                    var single = combos.Find(c => c.KeyCombination.Keys.Any(k => k == key))?.GetAction<T>();
+
+                    var blocked = HandleBindings?.GetInvocationList().Cast<Func<T?, List<T>, bool>>().ToList().FindAll(f => f.Invoke(single, pressedActions)).Any();
+
+                    if (blocked == true)
+                    {
+                        lastBlocked = e;
+                        return false;
+                    }
+
+                    lastBlocked = null;
+                }
+
+                return base.Handle(e);
+            }
+
+            public void RetractLastBlocked()
+            {
+                if (lastBlocked != null)
+                    base.Handle(lastBlocked);
+            }
```

Comment:
Suppressing an input just to retract that suppressing later feels backwards. This is too generic (as you need this only for the taiko implementation) and it shows.